<a href="https://colab.research.google.com/github/ilariacialfi/HousePricesAdvRegression/blob/main/ML_notebook_proj.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Regressione avanzata sui prezzi delle case
Il progetto prende spunto da una competizione pubblicata sul sito kaggle.com che ha l'obiettivo di predire il costo di una casa sulla base di 79 variabili che descrivono quasi ogni aspetto dell'immobile. Dovendo quindi predire un valore numerico e considerando di possedere un dataset contenente campioni di case con relativo prezzo di mercato quello che andiamo a fare è una regressione utilizzando algoritmi di apprendimento supervisionato. A tal fine verranno applicati e confrontati tra loro vari algoritmi:
* K-nearest neighbors
* Linear Discriminant Analysis
* SVM
* Random Forest
* Decision Tree
* Naive Bayes
* Logistic Regression
* 3-layer Neural Network
* AdaBoost

Il dataset utilizzato è "The Ames Housing" di Dean De Cock.
https://www.kaggle.com/c/house-prices-advanced-regression-techniques/overview

MOLTO UTILE
https://inria.github.io/scikit-learn-mooc/python_scripts/datasets_ames_housing.html


## Descrizione del dataset

Il dataset comprende due csv per training set e test set, un file di testo con la descrizione di ogni colonna e un ulteriore file csv che fa da benchmark dato da una regressione lineare su anno e mese di vendita, metratura del lotto e numero di camere da letto.

**SalePrice**: prezzo della 
proprietà. E' quindi la variabile target che cerchiamo di predire. E' posizionato come ultima colonna del csv.

Di seguito vediamo una breve descrizione degli attributi che verranno utilizzati:

1. **MSSubClass**: classe dell'abitazione
```
        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES
```
2. **MSZoning**: classificazione generale della zona
```
       A	Agriculture
       C	Commercial
       FV	Floating Village Residential
       I	Industrial
       RH	Residential High Density
       RL	Residential Low Density
       RP	Residential Low Density Park 
       RM	Residential Medium Density
```
3. **LotFrontage**: Lunghezza di strana che connette linearmente la proprietà (in piedi)

4. **LotArea**: Grandezza del lotto in piedi quadri

5. **Street**: Tipo di strada per accedere alla proprietà
```
       Grvl	Gravel	
       Pave	Paved
```
6. **Alley**: Tipo di vicolo per accedere alla proprietà
```
       Grvl	Gravel
       Pave	Paved
       NA 	No alley access
```
7. **LotShape**: Forma generale dell'immobile
```
       Reg	Regular	
       IR1	Slightly irregular
       IR2	Moderately Irregular
       IR3	Irregular
```
8. **LandContour**: Piattezza dell'immobile
```
       Lvl	Near Flat/Level	
       Bnk	Banked - Quick and significant rise from street grade to building
       HLS	Hillside - Significant slope from side to side
       Low	Depression
```
9. **Utilities**: Tipo di utenze disponibili
```
       AllPub	All public Utilities (E,G,W,& S)	
       NoSewr	Electricity, Gas, and Water (Septic Tank)
       NoSeWa	Electricity and Gas Only
       ELO	Electricity only	
```
10. **LotConfig**: Configurazione del lotto
```
       Inside	Inside lot
       Corner	Corner lot
       CulDSac	Cul-de-sac
       FR2	Frontage on 2 sides of property
       FR3	Frontage on 3 sides of property
```
11. **LandSlope**: Pendnza della proprietà
```
       Gtl	Gentle slope
       Mod	Moderate Slope	
       Sev	Severe Slope
```
12. **Neighborhood**: Posizioni fisiche entro i confini della città di Ames
```
       Blmngtn	Bloomington Heights
       Blueste	Bluestem
       BrDale	Briardale
       BrkSide	Brookside
       ClearCr	Clear Creek
       CollgCr	College Creek
       Crawfor	Crawford
       Edwards	Edwards
       Gilbert	Gilbert
       IDOTRR	Iowa DOT and Rail Road
       MeadowV	Meadow Village
       Mitchel	Mitchell
       Names	North Ames
       NoRidge	Northridge
       NPkVill	Northpark Villa
       NridgHt	Northridge Heights
       NWAmes	Northwest Ames
       OldTown	Old Town
       SWISU	South & West of Iowa State University
       Sawyer	Sawyer
       SawyerW	Sawyer West
       Somerst	Somerset
       StoneBr	Stone Brook
       Timber	Timberland
       Veenker	Veenker
```
13. **Condition1**: Vicinanza a diverse condizioni
```
       Artery	Adjacent to arterial street
       Feedr	Adjacent to feeder street	
       Norm	Normal	
       RRNn	Within 200' of North-South Railroad
       RRAn	Adjacent to North-South Railroad
       PosN	Near positive off-site feature--park, greenbelt, etc.
       PosA	Adjacent to postive off-site feature
       RRNe	Within 200' of East-West Railroad
       RRAe	Adjacent to East-West Railroad
```
14. **Condition2**: Vicinanza a diverse condizioni (se ne è preesente più di una)
```
       Artery	Adjacent to arterial street
       Feedr	Adjacent to feeder street	
       Norm	Normal	
       RRNn	Within 200' of North-South Railroad
       RRAn	Adjacent to North-South Railroad
       PosN	Near positive off-site feature--park, greenbelt, etc.
       PosA	Adjacent to postive off-site feature
       RRNe	Within 200' of East-West Railroad
       RRAe	Adjacent to East-West Railroad
```
15. **BldgType**: Tipo di abitazione
```
       1Fam	Single-family Detached	
       2FmCon	Two-family Conversion; originally built as one-family dwelling
       Duplx	Duplex
       TwnhsE	Townhouse End Unit
       TwnhsI	Townhouse Inside Unit
```
16. **HouseStyle**: Stile dell'abitazione
```
       1Story	One story
       1.5Fin	One and one-half story: 2nd level finished
       1.5Unf	One and one-half story: 2nd level unfinished
       2Story	Two story
       2.5Fin	Two and one-half story: 2nd level finished
       2.5Unf	Two and one-half story: 2nd level unfinished
       SFoyer	Split Foyer
       SLvl	Split Level
```
17. **OverallQual**: Valutazione del materiale complessivo e della finitura della casa
```
       10	Very Excellent
       9	Excellent
       8	Very Good
       7	Good
       6	Above Average
       5	Average
       4	Below Average
       3	Fair
       2	Poor
       1	Very Poor
```
18. **OverallCond**: Valutazione delle condizioni generali della casa
```
       10	Very Excellent
       9	Excellent
       8	Very Good
       7	Good
       6	Above Average	
       5	Average
       4	Below Average	
       3	Fair
       2	Poor
       1	Very Poor
```
19. **YearBuilt**: Data di costruzione originaria

20. **YearRemodAdd**: Data di ristrutturazion (stessa data della costruzione se non è stata ristrutturata)

21. **RoofStyle**: Tipo di tetto
```
       Flat	Flat
       Gable	Gable
       Gambrel	Gabrel (Barn)
       Hip	Hip
       Mansard	Mansard
       Shed	Shed
```
22. **RoofMatl**: Materiale del tetto
```
       ClyTile	Clay or Tile
       CompShg	Standard (Composite) Shingle
       Membran	Membrane
       Metal	Metal
       Roll	Roll
       Tar&Grv	Gravel & Tar
       WdShake	Wood Shakes
       WdShngl	Wood Shingles
```
23. **Exterior1st**: Rivestimento esterno della casa
```
       AsbShng	Asbestos Shingles
       AsphShn	Asphalt Shingles
       BrkComm	Brick Common
       BrkFace	Brick Face
       CBlock	Cinder Block
       CemntBd	Cement Board
       HdBoard	Hard Board
       ImStucc	Imitation Stucco
       MetalSd	Metal Siding
       Other	Other
       Plywood	Plywood
       PreCast	PreCast	
       Stone	Stone
       Stucco	Stucco
       VinylSd	Vinyl Siding
       Wd Sdng	Wood Siding
       WdShing	Wood Shingles
```
24. **Exterior2nd**: Rivestimento esterno della casa (se è stato utilizzato più di un materiale) 
```
       AsbShng	Asbestos Shingles
       AsphShn	Asphalt Shingles
       BrkComm	Brick Common
       BrkFace	Brick Face
       CBlock	Cinder Block
       CemntBd	Cement Board
       HdBoard	Hard Board
       ImStucc	Imitation Stucco
       MetalSd	Metal Siding
       Other	Other
       Plywood	Plywood
       PreCast	PreCast
       Stone	Stone
       Stucco	Stucco
       VinylSd	Vinyl Siding
       Wd Sdng	Wood Siding
       WdShing	Wood Shingles
```
25. **MasVnrType**: Tipo di rivestimento in mutatura
```
       BrkCmn	Brick Common
       BrkFace	Brick Face
       CBlock	Cinder Block
       None	None
       Stone	Stone
```
26. **MasVnrArea**: Superficie in muratura in piedi quadri

27. **ExterQual**: Valutazione della qualità dei materiali per l'esterno
```
       Ex	Excellent
       Gd	Good
       TA	Average/Typical
       Fa	Fair
       Po	Poor
```
28. **ExterCond**: Valutazione delle presenti condizioni del materiale esterno
```
       Ex	Excellent
       Gd	Good
       TA	Average/Typical
       Fa	Fair
       Po	Poor
```
29. **Foundation**: Tipo di fondamenta
```
       BrkTil	Brick & Tile
       CBlock	Cinder Block
       PConc	Poured Contrete	
       Slab	Slab
       Stone	Stone
       Wood	Wood
```
30. **BsmtQual**: Valutazione dell'altezza del seminterrato
```
       Ex	Excellent (100+ inches)	
       Gd	Good (90-99 inches)
       TA	Typical (80-89 inches)
       Fa	Fair (70-79 inches)
       Po	Poor (< 70 inches)
       NA	No Basement
```
31. **BsmtCond**: Valutazione delle condizioni generali del seminterrato
```
       Ex	Excellent
       Gd	Good
       TA	Typical - slight dampness allowed
       Fa	Fair - dampness or some cracking or settling
       Po	Poor - Severe cracking, settling, or wetness
       NA	No Basement
```
32. **BsmtExposure**: In riferimento ai livelli delle mura di recinzione 
```
       Gd	Good Exposure
       Av	Average Exposure (split levels or foyers typically score average or above)	
       Mn	Mimimum Exposure
       No	No Exposure
       NA	No Basement
```
33. **BsmtFinType1**: Valutazione dell'area finita del seminterrato
```
       GLQ	Good Living Quarters
       ALQ	Average Living Quarters
       BLQ	Below Average Living Quarters	
       Rec	Average Rec Room
       LwQ	Low Quality
       Unf	Unfinshed
       NA	No Basement
```
34. **BsmtFinSF1**: Piedi quadrati dell'area Type 1 finita

35. **BsmtFinType2**: Valutazione dell'area finita del seminterrato (se ci sono più tipologie)
```
       GLQ	Good Living Quarters
       ALQ	Average Living Quarters
       BLQ	Below Average Living Quarters	
       Rec	Average Rec Room
       LwQ	Low Quality
       Unf	Unfinshed
       NA	No Basement
```
36. **BsmtFinSF2**: Piedi quadrati dell'area Type 2 finita

37. **BsmtUnfSF**: Piedi quadrati dell'area del seminterrato non finita

38. **TotalBsmtSF**: Piedi quadrati totali dell'area del seminterrato

39. **Heating**: Tipo di riscaldamento
```
       Floor	Floor Furnace
       GasA	Gas forced warm air furnace
       GasW	Gas hot water or steam heat
       Grav	Gravity furnace	
       OthW	Hot water or steam heat other than gas
       Wall	Wall furnace
```
40. **HeatingQC**: Qualità e stato del riscaldamento
```
       Ex	Excellent
       Gd	Good
       TA	Average/Typical
       Fa	Fair
       Po	Poor
```
41. **CentralAir**: Aria condizionata centralizzata
```
       N	No
       Y	Yes
```
42. **Electrical**: Sistema elettrico
```
       SBrkr	Standard Circuit Breakers & Romex
       FuseA	Fuse Box over 60 AMP and all Romex wiring (Average)	
       FuseF	60 AMP Fuse Box and mostly Romex wiring (Fair)
       FuseP	60 AMP Fuse Box and mostly knob & tube wiring (poor)
       Mix	Mixed
```
43. **1stFlrSF**: piedi quadrati del primo piano
 
44. **2ndFlrSF**: piedi quadrati del secondo piano

45. **LowQualFinSF**: Piedi quadrati rifiniti di bassa qualità (tutti i piani)

46. **GrLivArea**: Superficie abitabile sopraelevata (terra) piedi quadrati

47. **BsmtFullBath**:Bagni completi seminterrato

48. **BsmtHalfBath**: Bagni a metà seminterrato

49. **FullBath**: Bagni completi a livello superiore

50. **HalfBath**: Bagni a metà a livello superiore

51. **Bedroom**: Camere da letto ai piani superiori (NON include le camere nel seminterrato)

52. **Kitchen**: Cucine ai piani superiori

53. **KitchenQual**: Qualità della cucina
```
       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor
```
54. **TotRmsAbvGrd**: Numero totale di stanze ai piani superiori (non include i bagni)

55. **Functional**: Funzionalità della casa (Assume typical unless deductions are warranted)
```
       Typ	Typical Functionality
       Min1	Minor Deductions 1
       Min2	Minor Deductions 2
       Mod	Moderate Deductions
       Maj1	Major Deductions 1
       Maj2	Major Deductions 2
       Sev	Severely Damaged
       Sal	Salvage only
```
56. **Fireplaces**: Numero di camini

57. **FireplaceQu**: Qualità dei camini
```
       Ex	Excellent - Exceptional Masonry Fireplace
       Gd	Good - Masonry Fireplace in main level
       TA	Average - Prefabricated Fireplace in main living area or Masonry Fireplace in basement
       Fa	Fair - Prefabricated Fireplace in basement
       Po	Poor - Ben Franklin Stove
       NA	No Fireplace
```
58. **GarageType**: Posizione del garage
```
       2Types	More than one type of garage
       Attchd	Attached to home
       Basment	Basement Garage
       BuiltIn	Built-In (Garage part of house - typically has room above garage)
       CarPort	Car Port
       Detchd	Detached from home
       NA	No Garage
```
59. **GarageYrBlt**: Anno di costruzione del garage
		
60. **GarageFinish**: Finitura interiore del garage
```
       Fin	Finished
       RFn	Rough Finished	
       Unf	Unfinished
       NA	No Garage
```
61. **GarageCars**: Grandezza del garage in capacità di automobili

62. **GarageArea**: Grandezza del garage in piedi quadri

63. **GarageQual**: Qualità del garage
```
       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor
       NA	No Garage
```
64. **GarageCond**: Condizione del garage
```
       Ex	Excellent
       Gd	Good
       TA	Typical/Average
       Fa	Fair
       Po	Poor
       NA	No Garage
```
65. **PavedDrive**: Vialetto lastricato
```
       Y	Paved 
       P	Partial Pavement
       N	Dirt/Gravel
```
66. **WoodDeckSF**: Area coperta in legno in piedi quadrati

67. **OpenPorchSF**: Area porticata aperta in piedi quadrati

68. **EnclosedPorch**: Zone porticata recintata in piedi quadrati

69. **3SsnPorch**: Portico tristagionale in piedi quadrati

70. **ScreenPorch**: Schermo del portico in metri quadrati

71. **PoolArea**: Piscina in piedi quadrati

72. **PoolQC**: Qualità della piscina
```
       Ex	Excellent
       Gd	Good
       TA	Average/Typical
       Fa	Fair
       NA	No Pool
```
73. **Fence**: Qualità del recinto
```
       GdPrv	Good Privacy
       MnPrv	Minimum Privacy
       GdWo	Good Wood
       MnWw	Minimum Wood/Wire
       NA	No Fence
```
74. **MiscFeature**: Funzionalità varie non coperte in altre categorie
```
       Elev	Elevator
       Gar2	2nd Garage (if not described in garage section)
       Othr	Other
       Shed	Shed (over 100 SF)
       TenC	Tennis Court
       NA	None
```
75. **MiscVal**: $Valore delle caratteristiche varie

76. **MoSold**: Mese della vendita (MM)

77. **YrSold**: Anno della vendita (YYYY)

78. **SaleType**: Tipo di vendita
```
       WD 	Warranty Deed - Conventional
       CWD	Warranty Deed - Cash
       VWD	Warranty Deed - VA Loan
       New	Home just constructed and sold
       COD	Court Officer Deed/Estate
       Con	Contract 15% Down payment regular terms
       ConLw	Contract Low Down payment and low interest
       ConLI	Contract Low Interest
       ConLD	Contract Low Down
       Oth	Other
```
79. **SaleCondition**: Condizioni di vendita
```
       Normal	Normal Sale
       Abnorml	Abnormal Sale -  trade, foreclosure, short sale
       AdjLand	Adjoining Land Purchase
       Alloca	Allocation - two linked properties with separate deeds, typically condo with a garage unit	
       Family	Sale between family members
       Partial	Home was not completed when last assessed (associated with New Homes)
```


## Processamento dei dati


###Importazione librerie

Importo le librerie che sono necessarie al processamento dei dati. In particolare utilizzerò:
* **numpy** necessaria per utilizzare funzioni matematiche in Python
* **pandas** che offre funzioni matematiche e di analisi dei dati
* **seaborn**, basato su matplotlib, che permette quindi di visualizzare grafici statistici
* **matplotlib** anch'esso per la visualizzazione di grafici
* **sklearn** con funzioni utili all'apprendimento automatico
* **tensorflow** anch'essa utile per l'apprendimento automatico e la geerazione di grafici
* **keras** utile per apprendimento automatico e reti neurali

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.layers import Flatten,Dense,Dropout,BatchNormalization,Conv2D,MaxPooling2D
from tensorflow.keras.models import Sequential
from keras import regularizers
from tensorflow.keras.metrics import Recall
from tensorflow.keras.callbacks import EarlyStopping,ReduceLROnPlateau,ModelCheckpoint
from sklearn.metrics import classification_report,confusion_matrix
from sklearn.preprocessing import LabelEncoder
from pandas import Series

###Importazione del dataset
Salvo il dataset su Google Drive e lo leggo

In [2]:
from google.colab import drive
drive.mount('/content/drive')

house_train = "/content/drive/MyDrive/house-prices-advanced-regression-techniques/train.csv"
house_test = "/content/drive/MyDrive/house-prices-advanced-regression-techniques/test.csv"

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
train = pd.read_csv(house_train)
train.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
5         Lvl    AllPub  ...        0    NaN  MnPrv        Shed     700   
6         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
7         Lvl    AllPub  ...        0    NaN    NaN        Shed     350   
8         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
9         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 81 columns]

In [7]:
print("Dimensionalità del dataset (righe):", train.shape[0])
print("Dimensione del dataset (colonne):", train.shape[1]-1)
print("Target : SalePrice")

Dimensionalità del dataset (righe): 1460
Dimensione del dataset (colonne): 80
Target : SalePrice


###Pulizia del dataset
Controllo se ci sono colonne che contengono valori nulli.

In [6]:
null_table= pd.DataFrame(train.isnull().sum().sort_values(ascending=False)/len(train))
null_table.columns=['null ratio']
null_table

null ratio
PoolQC         0.995205
MiscFeature    0.963014
Alley          0.937671
Fence          0.807534
FireplaceQu    0.472603
...                 ...
ExterQual      0.000000
Exterior2nd    0.000000
Exterior1st    0.000000
RoofMatl       0.000000
SalePrice      0.000000

[81 rows x 1 columns]

Andiamo quindi a sostituire i valori mancanti e controllo che non ce ne siano più.

In [8]:
train.fillna('')
for col in train.columns:
   train[col]= train[col].replace('','-')
   train[col]= train[col].replace(np.nan,'-')

null_table= pd.DataFrame(train.isnull().sum().sort_values(ascending=False)/len(train))
null_table.columns=['null ratio']
null_table

null ratio
Id                  0.0
CentralAir          0.0
GarageYrBlt         0.0
GarageType          0.0
FireplaceQu         0.0
...                 ...
MasVnrArea          0.0
MasVnrType          0.0
Exterior2nd         0.0
Exterior1st         0.0
SalePrice           0.0

[81 rows x 1 columns]

Abbiamo quindi sostituito tutti i valori mancanti.


###Label Encoder

Scikit-learn richiede che tutti i dati siano numerici, quindi è necessario convertire tutti i dati categotici.

Per prima cosa è quindi utile andare a vedere le informazioni circa il tipo di dato contenuto in ogni colonna.


In [14]:
train.info()
train.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Id             1460 non-null   int64 
 1   MSSubClass     1460 non-null   int64 
 2   MSZoning       1460 non-null   object
 3   LotFrontage    1460 non-null   object
 4   LotArea        1460 non-null   int64 
 5   Street         1460 non-null   object
 6   Alley          1460 non-null   object
 7   LotShape       1460 non-null   object
 8   LandContour    1460 non-null   object
 9   Utilities      1460 non-null   object
 10  LotConfig      1460 non-null   object
 11  LandSlope      1460 non-null   object
 12  Neighborhood   1460 non-null   object
 13  Condition1     1460 non-null   object
 14  Condition2     1460 non-null   object
 15  BldgType       1460 non-null   object
 16  HouseStyle     1460 non-null   object
 17  OverallQual    1460 non-null   int64 
 18  OverallCond    1460 non-null

Id  MSSubClass MSZoning LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL        65.0     8450   Pave     -      Reg   
1   2          20       RL        80.0     9600   Pave     -      Reg   
2   3          60       RL        68.0    11250   Pave     -      IR1   
3   4          70       RL        60.0     9550   Pave     -      IR1   
4   5          60       RL        84.0    14260   Pave     -      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0      -     -           -       0      2   
1         Lvl    AllPub  ...        0      -     -           -       0      5   
2         Lvl    AllPub  ...        0      -     -           -       0      9   
3         Lvl    AllPub  ...        0      -     -           -       0      2   
4         Lvl    AllPub  ...        0      -     -           -       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

Vedendo queste informazioni si evince che il dataset contiene 1460 campioni e che i dati che li descrivono sono sia numerici che categorici.
Andiamo quindi a modificare le colonne con valori categorici.


In [13]:
for column in train.columns:
    if column=='MSZoning' and column=='LotFrontage' and column=='Street' and column=='Alley' and column=='LotShape' and column=='LandContour' and column=='Utilities' and column=='LotConfig':
     train[column] = labelencoder.fit_transform(train[column])
train.head()

Id  MSSubClass MSZoning LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL        65.0     8450   Pave     -      Reg   
1   2          20       RL        80.0     9600   Pave     -      Reg   
2   3          60       RL        68.0    11250   Pave     -      IR1   
3   4          70       RL        60.0     9550   Pave     -      IR1   
4   5          60       RL        84.0    14260   Pave     -      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0      -     -           -       0      2   
1         Lvl    AllPub  ...        0      -     -           -       0      5   
2         Lvl    AllPub  ...        0      -     -           -       0      9   
3         Lvl    AllPub  ...        0      -     -           -       0      2   
4         Lvl    AllPub  ...        0      -     -           -       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [19]:
labelencoder=LabelEncoder()
cat_col = []
for col in train.columns:
  if train[col].dtypes!='int64':
    cat_col.append(col)
for column in train.columns:
    if column in cat_col:
     train[column] = labelencoder.fit_transform(train[column])

train.head(10)

TypeError: ignored

## Classificatori